File: tests/qt_types_standalone/rust/src/qlinef.rs
// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::{QLineF, QPointF};

#[cxx::bridge]
mod qline_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qlinef.h");
        type QLineF = cxx_qt_lib::QLineF;
    }

    extern "Rust" {
        fn construct_qlinef() -> QLineF;
        fn read_qlinef(m: &QLineF) -> bool;
        fn clone_qlinef(m: &QLineF) -> QLineF;
    }
}

fn construct_qlinef() -> QLineF {
    QLineF::new(QPointF::new(1.0, 2.0), QPointF::new(3.0, 4.0))
}

fn read_qlinef(m: &QLineF) -> bool {
    m.x1() == 1.0 && m.y1() == 2.0 && m.x2() == 3.0 && m.y2() == 4.0
}

fn clone_qlinef(m: &QLineF) -> QLineF {
    m.clone()
}

File: crates/cxx-qt-lib/src/core/qlinef.cpp
// clang-format off
// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Laurent Montel <laurent.montel@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qlinef.h"

#include "../assertion_utils.h"

#include <cstdint>

// QLineF has "QPointF" members - pt1, pt2
// Rust represents these as double types.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qline.h?h=v5.15.6-lts-lgpl#n281
//
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qline.h?h=v6.2.4#n295
constexpr static ::std::array<::std::size_t, 4> arr{ sizeof(double),
                                                     sizeof(double),
                                                     sizeof(double),
                                                     sizeof(double) };
assert_alignment_and_size(QLineF, alignof(double), arr);

static_assert(::std::is_trivially_copyable<QLineF>::value);
